From 3f33df4d9a338ae4af9ab438d050eb08a52eb029 Mon Sep 17 00:00:00 2001
From: Jacek Caban <jacek@codeweavers.com>
Date: Mon, 23 Feb 2015 12:14:49 +0100
Subject: [PATCH 1/3] Bug 1133689 - Make D3DVsyncDisplay destructor private.
 r=jmuizelaar


diff --git a/gfx/thebes/gfxWindowsPlatform.cpp b/gfx/thebes/gfxWindowsPlatform.cpp
index b52a188..23507ce 100644

Index: gfx/thebes/gfxWindowsPlatform.cpp
===================================================================
--- a/gfx/thebes/gfxWindowsPlatform.cpp
+++ b/gfx/thebes/gfxWindowsPlatform.cpp
@@ -1990,13 +1990,6 @@ public:
         mSoftwareVsyncRate = TimeDuration::FromMilliseconds(rate);
       }
 
-      virtual ~D3DVsyncDisplay()
-      {
-        MOZ_ASSERT(NS_IsMainThread());
-        DisableVsync();
-        delete mVsyncThread;
-      }
-
       virtual void EnableVsync() override
       {
         MOZ_ASSERT(NS_IsMainThread());
@@ -2099,6 +2092,13 @@ public:
       }
 
     private:
+      virtual ~D3DVsyncDisplay()
+      {
+        MOZ_ASSERT(NS_IsMainThread());
+        DisableVsync();
+        delete mVsyncThread;
+      }
+
       bool IsInVsyncThread()
       {
         return mVsyncThread->thread_id() == PlatformThread::CurrentId();
