From 7b6be41fce3486aa5b0deee676bd9d2242692476 Mon Sep 17 00:00:00 2001
From: Trevor Saunders <tbsaunde@tbsaunde.org>
Date: Fri, 20 Mar 2015 15:18:26 -0400
Subject: [PATCH 2/3] bug 1183967 - fixup correct case of mfidl.h


diff --git a/gfx/layers/IMFYCbCrImage.h b/gfx/layers/IMFYCbCrImage.h
index 5d863b4..6a5d392 100644

Index: gfx/layers/IMFYCbCrImage.h
===================================================================
--- a/gfx/layers/IMFYCbCrImage.h
+++ b/gfx/layers/IMFYCbCrImage.h
@@ -8,7 +8,7 @@
 
 #include "mozilla/RefPtr.h"
 #include "ImageContainer.h"
-#include "Mfidl.h"
+#include "mfidl.h"
 
 namespace mozilla {
 namespace layers {
