--- a/browser/components/newtab/lib/ActivityStream.jsm.orig	1980-01-01 18:59:51.000000000 -0500
+++ b/browser/components/newtab/lib/ActivityStream.jsm	2019-11-17 01:16:54.431504079 -0500
@@ -155,24 +155,8 @@
   }],
   ["improvesearch.topSiteSearchShortcuts.searchEngines", {
     title: "An ordered, comma-delimited list of search shortcuts that we should try and pin",
-    // This pref is dynamic as the shortcuts vary depending on the region
-    getValue: ({geo}) => {
-      if (!geo) {
-        return "";
-      }
-      const searchShortcuts = [];
-      if (geo === "CN") {
-        searchShortcuts.push("baidu");
-      } else if (["BY", "KZ", "RU", "TR"].includes(geo)) {
-        searchShortcuts.push("yandex");
-      } else {
-        searchShortcuts.push("google");
-      }
-      if (["DE", "FR", "GB", "IT", "JP", "US"].includes(geo)) {
-        searchShortcuts.push("amazon");
-      }
-      return searchShortcuts.join(",");
-    },
+    // This pref is dynamic as the shortcuts could vary depending on the region
+    getValue: ({geo}) => { return ""; },
   }],
   ["improvesearch.topSiteSearchShortcuts.havePinned", {
     title: "A comma-delimited list of search shortcuts that have previously been pinned",
